src/dshould/prettyprint.d

```diff
@@ -0,0 +1,385 @@
+{
+    auto textRange = text.quoted;
+    auto result = parse(textRange);
+
+    if (!textRange.empty)
+    {
+        return Nullable!Tree();
+    }
+
+    return result;
+}
+
+@("parse a paren expression to a tree")
+unittest
+{
+    import dshould : equal, should;
+
+    parse("Foo").should.equal(Tree("Foo").nullable);
+    parse(`"Foo"`).should.equal(Tree(`"Foo"`).nullable);
+    parse("Foo,").should.equal(Nullable!Tree());
+    parse("Foo, Bar").should.equal(Nullable!Tree());
+    parse("Foo()").should.equal(Tree("Foo", ParenType.paren.nullable));
+    parse("Foo[a, b]").should.equal(Tree(
+        "Foo",
+        ParenType.squareBracket.nullable,
+        [Tree("a"), Tree(" b")]).nullable);
+    parse(`Foo{"\""}`).should.equal(Tree(
+        "Foo",
+        ParenType.curlyBracket.nullable,
+        [Tree(`"\""`)]).nullable);
+}
+
+private Nullable!Tree parse(ref QuotedText textRange)
+{
+    auto parenStart = textRange.findAmong("({[");
+    auto closer = textRange.findAmong(",)}]");
+
+    if (textRange.textUntil(closer).length < textRange.textUntil(parenStart).length)
+    {
+        const prefix = textRange.textUntil(closer);
+
+        textRange = closer;
+        return Tree(prefix).nullable;
+    }
+
+    const prefix = textRange.textUntil(parenStart);
+
+    if (parenStart.empty)
+    {
+        textRange = parenStart;
+        return Tree(prefix).nullable;
+    }
+
+    const parenType = () {
+        switch (parenStart.front)
+        {
+            case '(':
+                return ParenType.paren;
+            case '[':
+                return ParenType.squareBracket;
+            case '{':
+                return ParenType.curlyBracket;
+            default:
+                assert(false);
+        }
+    }();
+
+    textRange = parenStart;
+    textRange.popFront;
+
+    Tree[] children = null;
+
+    while (true)
+    {
+        if (textRange.empty)
+        {
+            return Nullable!Tree();
+        }
+        if (textRange.front == parenType.closing)
+        {
+            // single child, quote only
+            const quoteChild = textRange.textUntil(textRange);
+
+            if (!quoteChild.empty)
+            {
+                children ~= Tree(quoteChild);
+            }
+
+            textRange.popFront;
+
+            return Tree(prefix, Nullable!ParenType(parenType), children).nullable;
+        }
+
+        if (!children.empty)
+        {
+            if (textRange.front != ',')
+            {
+                return Nullable!Tree();
+            }
+            textRange.popFront;
+        }
+
+        auto child = parse(textRange);
+
+        if (child.isNull)
+        {
+            return Nullable!Tree();
+        }
+        children ~= child;
+    }
+}
+
+// prefix
+// prefix { [child[, child]*]? }
+// prefix ( [child[, child]*]? )
+// prefix [ [child[, child]*]? ]
+private struct Tree
+{
+    string prefix;
+
+    Nullable!ParenType parenType = Nullable!ParenType();
```

Comment:
Usually, there is one `Tree` subtype per `ParenType` value.
But how about putting fields `opening` and `closing` into this `struct` so that the corresponding functions are no longer needed?

Reply:
That allows, at least in principle, the possibility of mismatched `opening` and `closing`.